velox/functions/sparksql/DateTimeFunctions.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/DateTimeFunctions.h"
+#include "velox/expression/DecodedArgs.h"
+#include "velox/expression/VectorFunction.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+std::optional<Timestamp> makeTimeStampFromDecodedArgs(
+    vector_size_t row,
+    DecodedVector* yearVector,
+    DecodedVector* monthVector,
+    DecodedVector* dayVector,
+    DecodedVector* hourVector,
+    DecodedVector* minuteVector,
+    DecodedVector* microsVector) {
+  // Check hour.
+  auto hour = hourVector->valueAt<int32_t>(row);
+  if (hour < 0 || hour > 24) {
+    return std::nullopt;
+  }
+  // Check miniute.
+  auto minute = minuteVector->valueAt<int32_t>(row);
+  if (minute < 0 || minute > 60) {
+    return std::nullopt;
+  }
+  // Check microseconds.
+  auto micros = microsVector->valueAt<int64_t>(row);
+  if (micros < 0) {
+    return std::nullopt;
+  }
+  auto seconds = micros / util::kMicrosPerSec;
+  if (seconds > 60 || (seconds == 60 && micros % util::kMicrosPerSec != 0)) {
+    return std::nullopt;
+  }
+
+  // year, month, day will be checked in utils::daysSinceEpochFromDate;
+  try {
+    auto daysSinceEpoch = util::daysSinceEpochFromDate(
+        yearVector->valueAt<int32_t>(row),
+        monthVector->valueAt<int32_t>(row),
+        dayVector->valueAt<int32_t>(row));
+    // micros has at most 8 digits (2 for seconds + 6 for microseconds),
+    // thus it's safe to cast micros from int64_t to int32_t.
+    auto localMicros = util::fromTime(hour, minute, 0, (int32_t)micros);
+    return util::fromDatetime(daysSinceEpoch, localMicros);
+  } catch (const VeloxUserError& e) {
+    return std::nullopt;
+  } catch (const std::exception&) {
+    throw;
+  }
+}
+
+void setTimestampOrNull(
+    int32_t row,
+    std::optional<Timestamp> timestamp,
+    int64_t tzID,
+    FlatVector<Timestamp>* result) {
+  if (timestamp.has_value()) {
+    (*timestamp).toGMT(tzID);
+    result->set(row, *timestamp);
+  } else {
+    result->setNull(row, true);
+  }
+}
+
+class MakeTimestampFunction : public exec::VectorFunction {
```

Comment:
Can this function be implemented as a simple function?

Reply:
The 6th parameter is of decimal type. I'm not sure it it's possible to implement it as a simple function.

Reply:
> The 6th parameter is of decimal type.

Wow... I didn't realize that. Would you update the documentation to state that clearly? Are there restrictions on precision and scale? I assume scale must be <= 3.

Reply:
@marin-ma Can you remind me why can't we implement functions with decimal inputs as simple functions? I wonder if it is worth looking into extending the framework to support such functions. Does Fuzzer support such functions? If not, we need to extend it. 

CC: @rui-mo @PHILO-HE @FelixYBW

Reply:
@mbasmanova Looks like the `resolver` used by simple fucntions doesn't support resolving decimal type https://github.com/facebookincubator/velox/blob/main/velox/expression/UdfTypeResolver.h
The expression fuzzer test doesn't use above framework so it can create decimal vectors.

Reply:
> The expression fuzzer test doesn't use above framework so it can create decimal vectors.

@marin-ma Are you saying that Fuzzer does cover this function? Would you run the Fuzzer with --only make_timestamp to make sure there are no failures?

Reply:
@mbasmanova I get exception with this command `./velox/expression/tests/spark_expression_fuzzer_test --only make_timestamp`

```
E0311 22:59:16.962633 2410940 Exceptions.h:69] Line: /home/sparkuser/github/oap-project/velox/velox/expression/tests/ExpressionFuzzer.cpp:1415, Function:fuzzReturnType, Expression: !signatures_.empty() No function signature available., Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: No function signature available.
Retriable: False
Expression: !signatures_.empty()
Function: fuzzReturnType
File: /home/sparkuser/github/oap-project/velox/velox/expression/tests/ExpressionFuzzer.cpp
```

Does it mean this function is not covered by the fuzzer test? Or did I use the wrong command?

Reply:
@marin-ma I believe Fuzzer doesn't support DECIMAL types yet. It would be nice to add this support, otherwise, test coverage of VectorFunctions that use DECIMAL types is limited.

CC: @rui-mo @PHILO-HE @majetideepak

Reply:
@mbasmanova Yes, fuzzer test for decimal type is not supported. This comment is about my previous finding https://github.com/facebookincubator/velox/pull/5791#issuecomment-1842440425, and I will remove the two limitations to see where the gap is, thanks.